### ISSUE TITLE
Mocking optional method parameters

### ISSUE THREAD

**Issue by reinink** [padraic/mockery#476]
Hey! So, I'm working through a pretty unique use case, but I was hoping for some help. I have a situation where I need to mock and test a method call, except in some situations there will be 1 argument, and in another situation, 2 arguments. This has to do with the use of a helper function. Consider the following:

## Example

```php
class MyClass
{
    public function myMethod($first, $second = null)
    {
    }
}

function my_helper($first, $second = null)
{
    $class = new MyClass();
    $class->myMethod($first, $second);
}
```

**Method called with one argument:**

```php
$class = new MyClass();
$class->myMethod('foo');
```

**Method called with two arguments:**

```php
my_helper('foo');
```

## Mocking

Here is what the mock would look like (roughly):

```php
$mock = Mockery::mock('MyClass');
$mock->shouldReceive('myMethod')
     ->atLeast()
     ->once()
     ->with(Mockery::type('string'));
```

## The problem

I'm in a situation where this test must account for either calling the `myMethod` method directly on a class instance, OR use the helper method. However, the helper method ALWAYS passes two arguments, where it's possible to only send one value when calling the method directly. **This basically means it's impossible to create a mock that satisfies both situations.**

I have tried everything, and it doesn't seem like there is anyway to get around this. Something like this would be great:

```php
$mock = Mockery::mock('MyClass');
$mock->shouldReceive('myMethod')
     ->atLeast()
     ->once()
     ->with(Mockery::type('string'), Mockery::optional());
```

Do you know of any workaround for this issue?

**Comment by davedevelopment** [comment]
``` php
$mock = Mockery::mock('MyClass');
$mock->shouldReceive('myMethod')
     ->atLeast()
     ->once()
     ->andReturnUsing(function($arg1, $arg2 = null) {
        $this->assertInternalType('string', $arg1);
        if ($arg2 !== null) {
            $this->assert...;
        }
    });

```

Maybe get around it? Typed on phone, so be wary of typos :)

**Comment by reinink** [comment]
Nice! Thanks Dave!

That's definitely a step in the right direction. It would be really cool if `andReturnUsing()` could return `true` or `false`, depending on whether it meets a specific condition.

Right now I want to test for `->atLeast()->once()`. Is it possible to use `andReturnUsing()` with those methods?

**Comment by davedevelopment** [comment]
We can't do the return thing, as `andReturnUsing`'s main use case is to be able to to return values based on inputs, so `m::mock()->shouldReceive("addOne")->with(m::type('int'))->andReturnUsing(function ($number) { return $number + 1; });`.

`andReturnUsing` should work fine with those methods, I can't think of a reason why not.

**Comment by reinink** [comment]
Ahh right, of course, I understand the purpose of `andReturnUsing` now. This method does work with `atLeast` and `once`, but my challenge is my particular test. It seems simple at first, but...

Call `myMethod` with the first argument having a class type of `foo` `atLeast` `once`.

Meaning `myMethod` could be called 5 other times with a different object as the first parameter, but as long as it's called once with `foo`, the test will succeed.

I think what I need to do is basically track each call to `myMethod`, and then do an assertion afterwards to ensure it was called at least once with `foo`.

**Comment by davedevelopment** [comment]
In that case you might be better off using the spy functionality, though the andReturnUsing doesn't work with that approach.

``` php

$spy = m::spy(MyClass::class);

// do stuff 

$spy->shouldHaveReceived('myMethod')->with(m::type(Foo::class)); 
```

**Comment by reinink** [comment]
Hmm, what's this now? What exactly are spies? I don't remember seeing anything about them in the docs.

**Comment by reinink** [comment]
So they are the same as mocked objects with `shouldIgnoreMissing()` enabled?

**Comment by davedevelopment** [comment]
Yes, they ignore calls but record them for later verification.

They're not documented yet (my fault, apologies), but I do have a blog post http://www.davedevelopment.co.uk/2014/10/09/mockery-spies.html

**Comment by reinink** [comment]
That's funny, my particular use case is with an event dispatcher. :)

Okay, so I still won't be able to use `with()`, because it will have the same issues that a regular mock had, but I can analyze the test better with the spy afterwards?

**Comment by reinink** [comment]
But differently, how do I get a list of the methods called, and the augments used from this spy?

**Comment by reinink** [comment]
It seems this data is behind protected methods, like `_mockery_getReceivedMethodCalls()`.

**Comment by davedevelopment** [comment]
Yes, we don't expose that information, only allow assertions against it. I'm not wholly opposed to exposing it, it just would have meant spending more time finalising more API for public consumption. So you're back to `andReturnUsing` and manual mock expectations maybe.

``` php

$mock = m::mock();
$expectedCallCount = 0;
$mock->shouldReceive('myMethod')
     ->andReturnUsing(function($arg1, $arg2 = null) use (&$expectedCallCount) {
        if ($arg1 instanceof Foo) {
            $expectedCallCount++;
        }
    });

// do stuff

if ($expectedCallCount < 1) {
    throw \Exception("myMethod was not called with a Foo at least once");
}
```

Boiler plate I know, but the only other alternative I can think of is refactor. The fact that we've back and forthed looking for a solution would suggest the tests are trying to tell you something.

**Comment by reinink** [comment]
Perfect, this is exactly the solution I just came to! And yes, I understand on the refactoring bit, but I'm actually going through all this effort to fix a bug in Laravel 5.1, which you see in this [method](https://github.com/laravel/framework/blob/5.1/src/Illuminate/Foundation/Testing/ApplicationTrait.php#L58). So while this should never happen in a regular application's tests, from a framework perspective this sort of work around actually makes sense.

The problem with Laravel 5.1 is that it allows two ways to trigger events. One being directly on the dispatcher, and the other via a helper method. But the argument count is different. I know Taylor is a fan of using the helper, which is probably why he hasn't run into this problem.

Whatever the case, you've been very helpful Dave. Thank you so much! I think I've got it from here.
